README.md
# jest-wrap <sup>[![Version Badge][2]][1]</sup>

[![Build Status][3]][4]
[![dependency status][5]][6]
[![dev dependency status][7]][8]
[![License][license-image]][license-url]
[![Downloads][downloads-image]][downloads-url]

[![npm badge][11]][1]

Fluent pluggable interface for easily wrapping `describe`, `it`, and `test` blocks in Jest tests.

## Example

```js
var wrap = require('jest-wrap');
var expect = require('chai').expect;

var mockWindow = {
	location: {
		href: 'test/url'
	}
};
wrap().withGlobal('window', () => mockWindow).describe('mocked window', function () {
	it('is mocked', function () {
		expect(window).to.equal(mockWindow);
	});

	it('has the right URL', function () {
		expect(window.location.href).to.equal('test/url');
	});
});

var obj = { a: 1 };
wrap().withOverrides(() => obj, () => ({ a: 2, b: 3 })).describe('overridden object keys', function () {
	it('has "b"', function () {
		expect(obj.b).to.equal(3);
	});

	it('has overridden "a"', function () {
		expect(obj.a).to.equal(2);
	});
});

wrap().withOverride(() => obj, 'a', () => 4).skip().describe('this test is skipped', function () {
	it('also supports .only()!', function () {
		expect(true).to.equal(false); // skipped
	});
});
```

## Plugins
A `jest-wrap` **plugin** is a named function that returns a JestWrapper instance or a descriptor object.
 - A plugin’s function `name` must begin with the string “with”.
 - Plugins can be globally registered, or `.use`d ad-hoc.
  - `.use` requires a plugin function as its first argument; further arguments are passed through to the plugin.
  - `.extend` requires a non-empty description string, and a descriptor object which may contain a value that is a function, or an array of functions, whose keys correspond to any or all of the supported jest methods.
 - Globally registered plugins, `.use` calls, and `.extend` calls can be chained, stored, and reused - each link in the chain creates a new instance of a JestWrapper.

 - A descriptor object may contain any or all of these 5 keys:
  - a `description` string, for use in “describe” and/or “it” (this is required when returning an object)
  - `beforeEach`: a function, or array of functions, for use in a `jest` `beforeEach` function
  - `afterEach`: a function, or array of functions, for use in a `jest` `afterEach` function
  - `beforeAll`: a function, or array of functions, for use in a `jest` `beforeAll` function
  - `afterAll`: a function, or array of functions, for use in a `jest` `afterAll` function

The most common approach will be for a plugin function to return `this.extend(description, descriptor)`.

A plugin function must have a `name` that starts with “with”, and will be invoked with a receiver (”this” value) of a JestWrapper instance.

To register a plugin, call the `register` function on `jest-wrap` with the plugin function. This should not be done in a reusable module.

```js
module.exports = function withFoo(any, args, you, want) {
	return this.extend('with some foo stuff', {
		beforeEach: function () {
			// setup ran before each test
		},
		afterEach: [
			function () {
				// teardown ran after each test
			},
			function () {
				// more teardown
			}
		],
		beforeAll: function () {
			// setup ran once before all tests
		},
		afterAll: function () {
			// teardown ran once after all tests
		}
	});
};
```

## Usage
```js
var wrap = require('jest-wrap');
wrap.register(require('jest-wrap-with-foo'));

wrap().withFoo().describe…
```

## skip/only
Although jest has `describe.skip`, `describe.only`, `it.skip`, `it.only`, `test.skip`, and `test.only`, it is not possible to implement these in jest-wrap without using ES5 property accessors. Since this project supports ES3, we decided to use `.skip().describe` etc rather than forfeit the ability to have skip/only.

## Tests
Simply clone the repo, `npm install`, and run `npm test`

[1]: https://npmjs.org/package/jest-wrap
[2]: http://versionbadg.es/ljharb/jest-wrap.svg
[3]: https://travis-ci.org/ljharb/jest-wrap.svg
[4]: https://travis-ci.org/ljharb/jest-wrap
[5]: https://david-dm.org/ljharb/jest-wrap.svg
[6]: https://david-dm.org/ljharb/jest-wrap
[7]: https://david-dm.org/ljharb/jest-wrap/dev-status.svg
[8]: https://david-dm.org/ljharb/jest-wrap#info=devDependencies
[11]: https://nodei.co/npm/jest-wrap.png?downloads=true&stars=true
[license-image]: http://img.shields.io/npm/l/jest-wrap.svg
[license-url]: LICENSE
[downloads-image]: http://img.shields.io/npm/dm/jest-wrap.svg
[downloads-url]: http://npm-stat.com/charts.html?package=jest-wrap


CHANGELOG.md
1.7.0 / 2020-05-12
==================
  * [New] adds support for Jest ^26

1.6.0 / 2020-01-22
==================
  * [New] adds support for Jest ^25

1.5.0 / 2019-02-11
==================
  * [New] adds support for Jest ^24

1.4.0 / 2018-06-08
==================
  * [New] adds support for Jest ^23

1.3.1 / 2018-05-01
=================
  * [Fix] ensure that skip works inside plugins, and plugins with no changes but the mode work
  * [Deps] update `semver`
  * [Dev Deps] update `eslint`, `istanbul-lib-coverage`, `nsp`, `tape`, `eslint-plugin-import`

1.3.0 / 2018-01-04
=================
  * [New] adds support for Jest ^22
  * [Deps] update `function.prototype.name`, `is-primitive`, `object-inspect`
  * [Dev Deps] update `eslint`, `eslint-config-airbnb-base`, `eslint-plugin-import`, `nsp`, `rimraf`
  * Move repo to airbnb

1.2.0 / 2017-05-12
=================
  * [New] add `jest` `v21` support
  * [Deps] update `function-bind`, `function.prototype.name`, `object-inspect`, `semver`
  * [Dev Deps] update `eslint`, `@ljharb/eslint-config`, `istanbul-lib-coverage`, `nsp`, `tape`
  * [Tests] make a matrix of jests
  * [Tests] only test major node versions; include `v8`
  * [Tests] use `nvm install-latest-npm` to ensure newer npms don’t break on older nodes

1.1.0 / 2017-05-12
=================
  * [New] Add jest 20 support (#9)
  * [Docs] Correct links/badges in the README (#8)
  * [Tests] on `node` `v7.10`
  * [Tests] Correct jest18/19 in package.json (#5)

1.0.2 / 2017-04-11
=================
  * [Fix] Fix descriptions when using multiple wrappers (#4)
  * [Fix] Stop reversing the afterEach hooks (#4)
  * [Fix] Update global beforeAll/afterAll hooks (#4)
  * [Fix] Remove .specify() (#4)
  * [Deps] update `object-inspect`
  * [Dev Deps] update `eslint`, `istanbul-lib-coverage`
  * [Tests] up to `node` `v7.9`
  * [Tests] Update core.js tests to work around bug in jest (#4)

1.0.1 / 2017-03-16
=================
  * [Fix] avoid exponentially adding outer wrappers (#3)
  * [Deps] lock `isarray` down to v1 only (v2 has a silly deprecation warning)
  * [Dev Deps] update `eslint`, `nsp`, `jest`, `rimraf`
  * [Tests] up to `node` `v7.7`, `v6.10`, `v4.8`

1.0.0 / 2017-02-18
=================
  * Initial release.


test/jest/core.js
'use strict';

var assert = require('assert');
var wrap = require('../../');

describe('core JestWrapper semantics', function () {
	describe('when there are no transformations', function () {
		it('throws when there are no transformations', function () {
			assert['throws'](function () { wrap().describe('foo', function () {}); }, RangeError);
			assert['throws'](function () { wrap().it('foo', function () {}); }, RangeError);
			assert['throws'](function () { wrap().test('foo', function () {}); }, RangeError);
		});

		it('does not throw when the mode is "skip"', function () {
			assert.doesNotThrow(function () { wrap().skip().describe('foo', function () {}); });
			assert.doesNotThrow(function () { wrap().skip().it('foo', function () {}); });
			assert.doesNotThrow(function () { wrap().skip().test('foo', function () {}); });
		});
	});

	var withNothing = function withNothing() {
		return { description: 'i am pointless' };
	};

	var testingCount = 0;
	var withTesting = function withTesting() {
		return this.extend('(with Testing)', {
			beforeEach: function withTestingBeforeEach() {
				testingCount += 1;
			}
		});
	};

	var withFancyNoop = function withFancyNoop() {
		return this.extend('(withFancyNoop)', {});
	};

	var withSkip = function withSkip() {
		return this.skip().extend('i am skipped', {});
	};

	wrap()
	.use(withSkip)
	.it('skips a test when the plugin skips it', function () {
		throw new SyntaxError('this should never run');
	});

	wrap()
	.use(withTesting)
	.use(withFancyNoop)
	.describe('with multiple plugins', function () {
		it('calls the plugin\'s hooks an appropriate number of times', function () {
			assert.equal(testingCount, 1);
		});
	});

	describe('#use()', function () {
		var flag = false;
		var withDescriptor = function withDescriptor() {
			return {
				description: 'i am a descriptor',
				beforeEach: function () { flag = true; },
				afterEach: function () { flag = false; }
			};
		};

		wrap().use(withDescriptor).describe('with a plugin that returns a descriptor', function () {
			it('works', function () {
				assert.equal(flag, true);
			});
		});

		wrap().use(withNothing).it('works with a plugin that returns a descriptor with only a description', function () {
			assert.equal(true, true); // oh yeah, TESTING
		});
	});

	describe('#skip()', function () {
		wrap().use(withNothing).skip().it('skipped an it!', function () {
			assert.equal(true, false); // boom
		});

		wrap().use(withNothing).skip().describe('skipped a describe!', function () {
			it('fails if not skipped', function () {
				assert.equal(true, false); // boom
			});
		});

		wrap().use(withNothing).skip().test('skipped a test!', function () {
			it('fails if not skipped', function () {
				assert.equal(true, false); // boom
			});
		});
	});

	/**
	 * Temporarily replace describe so that we can assert on the passed params.
	 */
	var originalDescribe = global.describe;
	var passedDescription;
	global.describe = function (description) {
		passedDescription = description;
		global.describe = originalDescribe; // revert to mocha's describe.
		return originalDescribe.apply(this, arguments);
	};

	wrap()
	.use(withFancyNoop)
	.use(withTesting)
	.describe('wrapped descriptions', function () {
		it('should have the proper description', function () {
			assert.equal(passedDescription, 'wrapped: (withFancyNoop); (with Testing):');
		});
	});

	var calls = [];
	describe('ordering of before/afters with multiple plugins', function () {
		var testsRun = 0;
		wrap().extend('first', {
			beforeAll: function () { calls.push('>:beforeAll'); },
			beforeEach: function () { calls.push('>:beforeEach'); },
			afterAll: function () { calls.push('>:afterAll'); },
			afterEach: function () { calls.push('>:afterEach'); }
		}).extend('second', {
			beforeAll: function () { calls.push('>>:beforeAll'); },
			beforeEach: function () { calls.push('>>:beforeEach'); },
			afterAll: function () { calls.push('>>:afterAll'); },
			afterEach: function () { calls.push('>>:afterEach'); }
		}).describe('with method tracking', function () {
			it('is one test', function () {
				testsRun += 1;
				calls.push('>>>:test');
			});

			it('is another test', function () {
				testsRun += 1;
				calls.push('>>>:test');
			});
		});

		it('runs the hooks/tests in the correct order', function () {
			if (testsRun !== 2) { throw new Error('This test cannot be run in isolation.'); }
			var expectedCalls = [
				'>:beforeAll',
				'>>:beforeAll',

				'>:beforeEach',
				'>>:beforeEach',
				'>>>:test',
				'>>:afterEach',
				'>:afterEach',

				'>:beforeEach',
				'>>:beforeEach',
				'>>>:test',
				'>>:afterEach',
				'>:afterEach',

				'>>:afterAll',
				'>:afterAll'
			];

			expect(calls).toEqual(expectedCalls);
		});
	});
});


test/tape/index.js
'use strict';

/* globals WeakMap */

var test = require('tape');

var wrap = require('../../');
var withGlobal = require('../../withGlobal');
var withOverrides = require('../../withOverrides');
var withOverride = require('../../withOverride');

var hasPrivacy = typeof WeakMap === 'function';

var setup = function setup() {
	var msgFn = function (msg, fn) { fn(); };
	global.describe = msgFn;
	global.test = msgFn;
	global.it = msgFn;
	var runFn = function (fn) { fn(); };
	global.before = runFn;
	global.beforeEach = runFn;
	global.after = runFn;
	global.afterEach = runFn;
};
var teardown = function teardown() {
	delete global.describe;
	delete global.test;
	delete global.it;
	delete global.before;
	delete global.beforeEach;
	delete global.after;
	delete global.afterEach;
};

test('jest-wrap', function (t) {
	setup();
	t.on('end', teardown);

	t.test('no transformations', function (st) {
		st['throws'](function () { wrap().describe('foo', function () {}); }, RangeError, 'throws when there are no transformations');
		st['throws'](function () { wrap().test('foo', function () {}); }, RangeError, 'throws when there are no transformations');
		st['throws'](function () { wrap().it('foo', function () {}); }, RangeError, 'throws when there are no transformations');
		st.end();
	});

	t.test('{describe,test,it}.{skip,only} throw', function (st) {
		st['throws'](function () { wrap().describe.skip('skip'); }, SyntaxError, 'describe.skip throws');
		st['throws'](function () { wrap().describe.only('only'); }, SyntaxError, 'describe.only throws');
		st['throws'](function () { wrap().test.skip('skip'); }, SyntaxError, 'test.skip throws');
		st['throws'](function () { wrap().test.only('only'); }, SyntaxError, 'test.only throws');
		st['throws'](function () { wrap().it.skip('skip'); }, SyntaxError, 'it.skip throws');
		st['throws'](function () { wrap().it.only('only'); }, SyntaxError, 'it.only throws');
		st.end();
	});

	t.test('withOverrides deferred exceptions', function (st) {
		st['throws'](function () {
			wrap().withOverrides(function () { return null; }, function () { return {}; }).describe('msg', function () {});
		}, TypeError, 'requires objectThunk to return an object');
		st['throws'](function () {
			wrap().withOverrides(function () { return {}; }, function () { return null; }).describe('msg', function () {});
		}, TypeError, 'requires overridesThunk to return an object');
		st.end();
	});

	t.test('.supportedMethods', function (st) {
		st.equal(Array.isArray(wrap.supportedMethods), true, 'is an array');
		st.deepEqual(wrap.supportedMethods, ['before', 'beforeEach', 'after', 'afterEach'], 'has the expected methods');
		st.end();
	});

	t.test('.register()', function (st) {
		st.equal(typeof wrap.register, 'function', 'is a function');

		st.test('throws on an invalid plugin name', function (st2) {
			st2['throws'](function () { wrap.register(); }, TypeError, 'throws on a non-string plugin name');
			st2['throws'](function () { wrap.register(''); }, TypeError, 'throws on an empty string plugin name');
			st2['throws'](function () { wrap.register('foo'); }, TypeError, 'throws on a non-prefixed plugin name');
			st2.end();
		});

		st.end();
	});

	t.test('.unregister()', function (st) {
		st.equal(typeof wrap.unregister, 'function', 'is a function');

		st.test('throws on an invalid plugin name', function (st2) {
			st2['throws'](function () { wrap.unregister(); }, TypeError, 'throws on a non-string plugin name');
			st2['throws'](function () { wrap.unregister(''); }, TypeError, 'throws on an empty string plugin name');
			st2['throws'](function () { wrap.unregister('foo'); }, TypeError, 'throws on a non-prefixed plugin name');
			st2.end();
		});

		st['throws'](function () { wrap.unregister('withNope'); }, RangeError, 'throws on an unregistered plugin');

		st.end();
	});

	var plugin = function withFoo() { return this.extend('…'); };
	t.test('registration/unregistration', function (st) {
		var instance = wrap();
		st.equal(typeof instance.withFoo, 'undefined', 'withFoo is not registered');
		wrap.register(plugin);
		st.equal(typeof instance.withFoo, 'function', 'withFoo is registered');
		wrap.unregister('withFoo');
		st.equal(typeof instance.withFoo, 'undefined', 'withFoo is unregistered by name');
		wrap.register(plugin);
		st.equal(typeof instance.withFoo, 'function', 'withFoo is again registered');
		wrap.unregister(plugin);
		st.equal(typeof instance.withFoo, 'undefined', 'withFoo is unregistered by function');
		st.end();
	});

	t.test('duplicate registration', function (st) {
		var instance = wrap();
		wrap.register(plugin);
		var original = instance.withFoo;
		wrap.register(plugin);
		st.equal(original, instance.withFoo, 'duplicate "register" is a noop');
		st.end();
	});

	t.end();
});

test('JestWrapper', function (t) {
	t.notEqual(wrap(), wrap(), 'wrap() returns different instances');

	var instance = wrap();
	var described = instance.extend('described!');
	t.notEqual(instance, described, 'sanity check: .extend() returns a new instance');

	t.test('hidden instance properties', { skip: !hasPrivacy }, function (st) {
		st.deepEqual(Object.keys(instance), [], 'has no own keys');
		st.deepEqual(Object.keys(described), [], 'has no own keys');

		st.end();
	});

	t.test('visible instance properties', { skip: hasPrivacy }, function (st) {
		st.deepEqual(Object.keys(instance), ['wrappers', 'mode'], 'has "wrappers" and "mode" key');
		st.deepEqual(Object.keys(described), ['wrappers', 'mode', 'description'], 'has "wrappers", "mode", and "description" keys');

		st.end();
	});

	t.test('#use()', function (st) {
		st.test('borrowing on a non-wrapper', function (sst) {
			sst['throws'](function () { instance.use.call({}, function withFoo() {}); }, TypeError, 'throws when receiver is not JestWrapper');
			sst.end();
		});

		st.test('non-functions', function (sst) {
			sst['throws'](function () { instance.use(); }, TypeError, 'throws with undefined');
			sst['throws'](function () { instance.use(null); }, TypeError, 'throws with null');
			sst['throws'](function () { instance.use(true); }, TypeError, 'throws with true');
			sst['throws'](function () { instance.use('foo'); }, TypeError, 'throws with string');
			sst['throws'](function () { instance.use(/a/g); }, TypeError, 'throws with regex');
			sst['throws'](function () { instance.use([]); }, TypeError, 'throws with array');
			sst['throws'](function () { instance.use({}); }, TypeError, 'throws with object');
			sst.end();
		});

		st.test('functions without the correct name', function (sst) {
			var anon = function () {};
			sst['throws'](function () { instance.use(anon); }, TypeError, 'throws with anonymous function');
			var badPlugin = function wrongPrefix() {};
			sst['throws'](function () { instance.use(badPlugin); }, TypeError, 'throws with badly named function');
			sst.end();
		});

		st.test('with a plugin', function (sst) {
			sst.test('plugins not returning a description', function (s2t) {
				s2t['throws'](
					function () { instance.use(function withFoo() {}); },
					TypeError,
					'throws when plugin returns undefined'
				);
				s2t['throws'](
					function () { instance.use(function withFoo() { return {}; }); },
					TypeError,
					'throws when plugin returns no description'
				);
				s2t['throws'](
					function () { instance.use(function withFoo() { return { description: true }; }); },
					TypeError,
					'throws when plugin returns nonstring description'
				);
				s2t['throws'](
					function () { instance.use(function withFoo() { return { description: '' }; }); },
					TypeError,
					'throws when plugin returns empty description'
				);
				s2t.end();
			});

			sst.test('receiver and proxied arguments', function (s2t) {
				s2t.plan(7);
				var plugin = function withFoo(a, b, c, d) {
					s2t.equal(arguments.length, 3, '3 args passed');
					s2t.equal(a, 1, 'a arg is 1');
					s2t.equal(b, 2, 'b arg is 1');
					s2t.equal(c, 3, 'c arg is 1');
					s2t.equal(d, undefined, 'd arg is undefined');
					s2t.equal(instance, this, 'plugin receiver is instance');
					return { description: 'something' };
				};
				var withPlugin = instance.use(plugin, 1, 2, 3);
				s2t.notEqual(instance, withPlugin, 'sanity check: "use" returns a new instance');
			});

			sst.test('plugin that calls .extend()', function (s2t) {
				var plugin = function withFoo() {
					return this.extend('description');
				};
				var withPlugin = instance.use(plugin);
				s2t.notEqual(instance, withPlugin, 'sanity check: "use" returns a new instance');
				s2t.end();
			});

			sst.end();
		});

		st.end();
	});

	t.test('#extend()', function (st) {
		st['throws'](function () { instance.extend(); }, TypeError, 'throws with no description');
		st['throws'](function () { instance.extend({}); }, TypeError, 'throws with non-string description');
		st['throws'](function () { instance.extend(''); }, TypeError, 'throws with empty description');

		var noDescriptor = instance.extend('…');
		st.notEqual(instance, noDescriptor, 'sanity check: "extend" returns a new instance with no descriptor');

		var emptyDescriptor = instance.extend('…', {});
		st.notEqual(instance, emptyDescriptor, 'sanity check: "extend" returns a new instance with empty descriptor');

		var extraDescriptor = instance.extend('…', { foo: 'bar' });
		st.notEqual(instance, extraDescriptor, 'sanity check: "extend" returns a new instance with descriptor with extra properties');

		st['throws'](function () { instance.extend('…', { before: true }); }, TypeError, 'throws with method override: noncallable function');
		st['throws'](function () { instance.extend('…', { before: [true] }); }, TypeError, 'throws with method override: array with noncallable function');

		var validDescriptor = instance.extend('…', { before: function () {}, beforeEach: [], afterEach: [function () {}] });
		st.notEqual(instance, validDescriptor, 'sanity check: "extend" returns a new instance with valid descriptor');

		st.end();
	});

	t.end();
});

test('withGlobal', function (t) {
	t.test('key exceptions', function (st) {
		st['throws'](withGlobal, TypeError, 'requires a string or Symbol');
		st['throws'](function () { withGlobal(''); }, TypeError, 'string must not be empty');
		st['throws'](function () { withGlobal(); }, TypeError, 'undefined is not a string or Symbol');
		st['throws'](function () { withGlobal(null); }, TypeError, 'null is not a string or Symbol');
		st['throws'](function () { withGlobal(true); }, TypeError, 'true is not a string or Symbol');
		st['throws'](function () { withGlobal(42); }, TypeError, '42 is not a string or Symbol');
		st['throws'](function () { withGlobal(/a/g); }, TypeError, 'regex is not a string or Symbol');
		st['throws'](function () { withGlobal([]); }, TypeError, 'array is not a string or Symbol');
		st['throws'](function () { withGlobal({}); }, TypeError, 'object is not a string or Symbol');
		st.end();
	});

	t.test('value exceptions', function (st) {
		st['throws'](function () { withGlobal('key', {}); }, TypeError, 'requires a callable valueThunk');
		st.end();
	});

	t.end();
});

test('withOverrides', function (t) {
	t.test('exceptions', function (st) {
		st['throws'](withOverrides, TypeError, 'requires an objectThunk to override');
		st['throws'](function () { withOverrides({}); }, TypeError, 'requires a callable objectThunk to override');
		st['throws'](function () { withOverrides(function () { return {}; }); }, TypeError, 'requires an objectThunk to override with');
		st.end();
	});

	t.test('works with functions', function (st) {
		st.doesNotThrow(function () {
			wrap().withOverrides(
				function () { return function () {}; },
				function () { return {}; }
			);
		}, 'accepts a function to override');
		st.doesNotThrow(function () {
			wrap().withOverrides(
				function () { return {}; },
				function () { return function () {}; }
			);
		}, 'accepts a function of overrides');
		st.end();
	});

	t.end();
});

test('withOverride', function (t) {
	t.test('exceptions', function (st) {
		var getObj = function () { return {}; };
		st['throws'](function () { withOverride(null, 'foo'); }, TypeError, 'requires an objectThunk');
		st['throws'](function () { withOverride({}, 'foo'); }, TypeError, 'requires a callable objectThunk');
		st['throws'](function () { withOverride(getObj); }, TypeError, 'undefined is not a string or Symbol');
		st['throws'](function () { withOverride(getObj, null); }, TypeError, 'null is not a string or Symbol');
		st['throws'](function () { withOverride(getObj, true); }, TypeError, 'true is not a string or Symbol');
		st['throws'](function () { withOverride(getObj, 42); }, TypeError, '42 is not a string or Symbol');
		st['throws'](function () { withOverride(getObj, /a/g); }, TypeError, 'regex is not a string or Symbol');
		st['throws'](function () { withOverride(getObj, []); }, TypeError, 'array is not a string or Symbol');
		st['throws'](function () { withOverride(getObj, {}); }, TypeError, 'object is not a string or Symbol');

		st['throws'](function () { withOverride(getObj, 'key', {}); }, TypeError, 'requires a callable valueThunk');
		st.end();
	});

	t.end();
});


test/jest-only/index.js
'use strict';

var assert = require('assert');
var wrap = require('../../');

var withNothing = function withNothing() {
	return this.extend('with nothing', { beforeAll: function () {} });
};

describe('#only()', function () {
	it('fails', function () {
		assert.equal(true, false, 'explode!');
	});

	wrap().use(withNothing).only().it('passes', function () {
		assert.equal(true, true, 'testing is fun');
	});
});


index.js
'use strict';

/* globals WeakMap */

var isCallable = require('is-callable');
var isString = require('is-string');
var has = require('has');
var forEach = require('for-each');
var isArray = require('isarray');
var functionName = require('function.prototype.name');
var inspect = require('object-inspect');
var semver = require('semver');
var jestVersion = require('jest').getVersion();

var checkWithName = require('./helpers/checkWithName');

var withOverrides = require('./withOverrides');
var withOverride = require('./withOverride');
var withGlobal = require('./withGlobal');

var hasPrivacy = typeof WeakMap === 'function';
var wrapperMap = hasPrivacy ? new WeakMap() : /* istanbul ignore next */ null;
var modeMap = hasPrivacy ? new WeakMap() : /* istanbul ignore next */ null;

var MODE_ALL = 'all';
var MODE_SKIP = 'skip';
var MODE_ONLY = 'only';

var beforeMethods = ['beforeAll', 'beforeEach'];
var afterMethods = ['afterAll', 'afterEach'];
var supportedMethods = [].concat(beforeMethods, afterMethods);

/**
 * There is a bug in Jest 18/19 that processes the afterAll hooks in the wrong
 * order. This bit of logic is meant to understand which method Jest is using
 * for moving through the list of afterAll hooks, and supply them in the order
 * that gets them applied in the order we want.
 */
var needsAfterAllReversal = semver.satisfies(jestVersion, '< 20');

var JestWrapper;

var checkThis = function requireJestWrapper(instance) {
	if (!instance || typeof instance !== 'object' || !(instance instanceof JestWrapper)) {
		throw new TypeError(inspect(instance) + ' must be a JestWrapper');
	}
	return instance;
};

var setThisWrappers = function (instance, value) {
	checkThis(instance);
	/* istanbul ignore else */
	if (hasPrivacy) {
		wrapperMap.set(instance, value);
	} else {
		instance.wrappers = value; // eslint-disable-line no-param-reassign
	}
	return instance;
};

var getThisWrappers = function (instance) {
	checkThis(instance);
	return hasPrivacy ? wrapperMap.get(instance) : /* istanbul ignore next */ instance.wrappers;
};

var setThisMode = function (instance, mode) {
	checkThis(instance);
	/* istanbul ignore else */
	if (hasPrivacy) {
		modeMap.set(instance, mode);
	} else {
		instance.mode = mode; // eslint-disable-line no-param-reassign
	}
	return instance;
};

var getThisMode = function (instance) {
	checkThis(instance);
	return hasPrivacy ? modeMap.get(instance) : /* istanbul ignore next */ instance.mode;
};

JestWrapper = function JestWrapper() { // eslint-disable-line no-shadow
	setThisWrappers(this, []);
	setThisMode(this, MODE_ALL);
};

var createWithWrappers = function (wrappers) {
	return setThisWrappers(new JestWrapper(), wrappers);
};

var concatThis = function (instance, toConcat) {
	var thisWrappers = getThisWrappers(instance);
	var thisMode = getThisMode(instance);
	return setThisMode(createWithWrappers(thisWrappers.concat(toConcat || [])), thisMode);
};

var flattenToDescriptors = function flattenToDescriptors(wrappers) {
	if (wrappers.length === 0) { return []; }

	var descriptors = [];
	forEach(wrappers, function (wrapper) {
		var subWrappers = wrapper instanceof JestWrapper ? getThisWrappers(wrapper) : wrapper;
		if (Array.isArray(subWrappers)) {
			descriptors.push.apply(descriptors, flattenToDescriptors(subWrappers));
		} else {
			descriptors.push(subWrappers);
		}
	});
	return descriptors;
};

var applyMethods = function applyMethods(methodsToApply, descriptors) {
	forEach(descriptors, function (methods) {
		forEach(methodsToApply, function (method) {
			var functions = methods[method];
			if (functions) {
				forEach(functions, function (func) {
					global[method](func);
				});
			}
		});
	});
};

var createAssertion = function createAssertion(type, message, wrappers, block, mode) {
	var descriptors = flattenToDescriptors(wrappers);
	if (descriptors.length === 0 && mode === MODE_ALL) {
		throw new RangeError(inspect(type) + ' called with no wrappers defined');
	}

	var describeMsgs = [];
	forEach(descriptors, function (descriptor) {
		if (descriptor.description) {
			describeMsgs.push(descriptor.description);
		}
	});

	var describeMsg = 'wrapped: ' + describeMsgs.join('; ') + ':';
	var describeMethod = global.describe;
	if (mode === MODE_SKIP) {
		describeMethod = global.describe.skip;
	} else if (mode === MODE_ONLY) {
		describeMethod = global.describe.only;
	}

	describeMethod(describeMsg, function () {
		applyMethods(beforeMethods, descriptors);
		global[type](message, block);

		// See comment at top of file.
		if (needsAfterAllReversal) {
			applyMethods(['afterEach'], descriptors);
			applyMethods(['afterAll'], descriptors.reverse());
		} else {
			applyMethods(afterMethods, descriptors);
		}
	});
};

JestWrapper.prototype.skip = function skip() {
	return setThisMode(concatThis(this), MODE_SKIP);
};

JestWrapper.prototype.only = function only() {
	return setThisMode(concatThis(this), MODE_ONLY);
};

JestWrapper.prototype.it = function it(msg, fn) {
	var wrappers = getThisWrappers(checkThis(this));
	var mode = getThisMode(this);
	createAssertion('it', msg, wrappers, fn, mode);
};
JestWrapper.prototype.it.skip = function skip() {
	throw new SyntaxError('jest-wrap requires `.skip().it` rather than `it.skip`');
};
JestWrapper.prototype.it.only = function only() {
	throw new SyntaxError('jest-wrap requires `.only().it` rather than `it.only`');
};

JestWrapper.prototype.test = function test(msg, fn) {
	var wrappers = getThisWrappers(checkThis(this));
	var mode = getThisMode(this);
	createAssertion('test', msg, wrappers, fn, mode);
};
JestWrapper.prototype.test.skip = function skip() {
	throw new SyntaxError('jest-wrap requires `.skip().test` rather than `test.skip`');
};
JestWrapper.prototype.test.only = function only() {
	throw new SyntaxError('jest-wrap requires `.only().test` rather than `test.only`');
};

JestWrapper.prototype.describe = function describe(msg, fn) {
	var wrappers = getThisWrappers(checkThis(this));
	var mode = getThisMode(this);
	createAssertion('describe', msg, wrappers, fn, mode);
};
JestWrapper.prototype.describe.skip = function skip() {
	throw new SyntaxError('jest-wrap requires `.skip().describe` rather than `describe.skip`');
};
JestWrapper.prototype.describe.only = function only() {
	throw new SyntaxError('jest-wrap requires `.only().describe` rather than `describe.only`');
};

var wrap = function wrap() { return new JestWrapper(); };

var isWithNameAvailable = function (name) {
	checkWithName(name);
	return !has(JestWrapper.prototype, name) || !isCallable(JestWrapper.prototype[name]);
};

wrap.supportedMethods = isCallable(Object.freeze)
	? Object.freeze(supportedMethods)
	: /* istanbul ignore next */ supportedMethods.slice();

JestWrapper.prototype.extend = function extend(description, descriptor) {
	checkThis(this);
	if (!isString(description) || description.length === 0) {
		throw new TypeError('a non-empty description string is required');
	}
	var newWrappers = [];
	if (descriptor) {
		forEach(supportedMethods, function (methodName) {
			if (methodName in descriptor) {
				if (!isArray(descriptor[methodName])) {
					// eslint-disable-next-line no-param-reassign
					descriptor[methodName] = [descriptor[methodName]];
				}
				forEach(descriptor[methodName], function (method) {
					if (!isCallable(method)) {
						throw new TypeError('wrapper method "' + method + '" must be a function, or array of functions, if present');
					}
				});
			}
		});
		descriptor.description = description; // eslint-disable-line no-param-reassign
		newWrappers = [createWithWrappers([descriptor])];
	}
	return concatThis(this, newWrappers);
};

JestWrapper.prototype.use = function use(plugin) {
	checkThis(this);
	if (!isCallable(plugin)) {
		throw new TypeError('plugin must be a function');
	}
	var withName = functionName(plugin);
	checkWithName(withName);

	var extraArguments = Array.prototype.slice.call(arguments, 1);
	var descriptorOrInstance = plugin.apply(this, extraArguments) || {};

	var instance = descriptorOrInstance;
	if (!(descriptorOrInstance instanceof JestWrapper)) {
		instance = wrap().extend(descriptorOrInstance.description, descriptorOrInstance);
	}

	var thisMode = getThisMode(instance);
	return setThisMode(setThisWrappers(new JestWrapper(), [instance]), thisMode);
};

wrap.register = function register(plugin) {
	var withName = functionName(plugin);
	checkWithName(withName);
	if (!isWithNameAvailable(withName)) {
		// already registered
		return;
	}
	JestWrapper.prototype[withName] = function wrapper() {
		return this.use.apply(this, [plugin].concat(Array.prototype.slice.call(arguments)));
	};
};

wrap.unregister = function unregister(pluginOrWithName) {
	var withName = isCallable(pluginOrWithName) ? functionName(pluginOrWithName) : pluginOrWithName;
	checkWithName(withName);
	if (isWithNameAvailable(withName)) {
		throw new RangeError('error: plugin "' + withName + '" is not registered.');
	}
	delete JestWrapper.prototype[withName];
};

wrap.register(withOverrides);
wrap.register(withOverride);
wrap.register(withGlobal);

module.exports = wrap;
